src/mlnet/Telemetry/Telemetry.cs

```diff
@@ -1,144 +1,239 @@
-            _trackEventTask = Task.Factory.StartNew(() => InitializeTelemetry());
+
+            // Initialize in task to offload to parallel thread
+            _initializationTask = Task.Factory.StartNew(() => InitializeTask());
         }
 
-        public void TrackEvent(
-            string eventName, 
+        /// <summary>
+        /// Send telemetry event.
+        /// </summary>
+        public static void TrackEvent(
+            string eventName,
             IDictionary<string, string> properties,
-            IDictionary<string, double> measurements)
+            TimeSpan? duration = null,
+            Exception ex = null)
         {
-            if (!Enabled)
+            if (!_enabled)
             {
                 return;
             }
-
-            //continue task in existing parallel thread
-            _trackEventTask = _trackEventTask.ContinueWith(
-                x => TrackEventTask(eventName, properties, measurements)
-            );
+            // Increment count of outstanding telemetry events.
+            Interlocked.Increment(ref _outstandingTelemetryEventCount);
+            _initializationTask.ContinueWith(x => TrackEventTask(eventName, properties, duration, ex));
         }
 
-        public void ThreadBlockingTrackEvent(string eventName, IDictionary<string, string> properties, IDictionary<string, double> measurements)
+        /// <summary>
+        /// Wait for all outstanding telemetry tasks to complete.
+        /// </summary>
+        public static void Flush(TimeSpan timeout)
         {
-            if (!Enabled)
+            var sw = Stopwatch.StartNew();
+            while (sw.Elapsed < timeout)
             {
-                return;
+                if (_outstandingTelemetryEventCount == 0)
+                {
+                    break;
+                }
+                Task.Delay(200).Wait();
             }
+        }
 
-            TrackEventTask(eventName, properties, measurements);
+        /// <summary>
+        /// Get serialized pipeline to log. Be careful to exclude PII.
+        /// </summary>
+        public static string GetSanitizedPipelineStr(Pipeline pipeline)
+        {
+            if (pipeline?.Nodes == null)
+            {
+                return null;
+            }
+            var transformNodes = pipeline.Nodes.Where(n => n.NodeType == PipelineNodeType.Transform);
+            var trainerNode = pipeline.Nodes.FirstOrDefault(n => n.NodeType == PipelineNodeType.Trainer);
+            var sb = new StringBuilder();
+            foreach (var transformNode in transformNodes)
+            {
+                sb.Append(transformNode.Name);
+                sb.Append(",");
+            }
+            if (trainerNode != null)
+            {
+                sb.Append(trainerNode.Name);
+                sb.Append("{");
+                var serializedHyperparams = trainerNode.Properties
+                    .Where(p => SweepableParams.AllHyperparameterNames.Contains(p.Key))
```

Comment:
trying to understand the intent here..  why not log all of them?  defaults can change from dataset to dataset and also over time for hyperparams we don't explicitly set no?  which would be an interesting bit of info as well.

Reply:
Thinking out loud here too: if the user has any custom stuff in the pipeline, can we end up logging PII here?  Or I guess at this point there can't be any custom components so we should be Ok but I don't like the fact that this is here and when we implement a feature to prepend custom things to pipelines someone might not think about telemetry/PII

Reply:
'LabelColumnName', 'ExampleWeightColumnName' etc are trainer properties here. I erred hard by choosing only to log properties that were whitelisted in a file that only contained a bunch of hyperparameter names to err as hard away from logging PII as possible. let me know if there's a better approach